x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -209,25 +210,29 @@ static boolean tierNodesPresent(String singleTier, DiscoveryNodes nodes) {
     }
 
 
-    private static boolean allocationAllowed(OpType opType, String tierSetting, Set<DiscoveryNodeRole> roles) {
+    private static boolean allocationAllowed(String tierSetting, Set<DiscoveryNodeRole> roles) {
         String[] values = parseTierList(tierSetting);
```

Comment:
We actually don't need to parse this setting at all, because the caller of this function does:

```java
            Optional<String> tier = preferredTierFunction.apply(tierPreference, allocation.nodes());
            if (tier.isPresent()) {
                String tierName = tier.get();
                if (allocationAllowed(tierName, roles)) {
```

They `preferredTierFunction` returns the optional value of a *single* tier name.

So we can guarantee that in this function, `tierSetting` is a non-empty, single-valued string for a given tier (that also exists in the cluster). So, all callers will go through the `values.length == 1` branch (assuming they aren't a generic `data` role node)

Reply:
Ah, excellent! d4668a8c17c